gst/datarepo/gstdatareposrc.c

```diff
@@ -1334,23 +1283,138 @@ gst_data_repo_src_read_json_file (GstDataRepoSrc * src)
   src->total_samples = json_object_get_int_member (object, "total_samples");
   GST_INFO_OBJECT (src, "total_samples: %d", src->total_samples);
 
+  if (src->total_samples == 0)
+    goto error;
+
   g_free (contents);
   g_object_unref (parser);
-  gst_caps_unref (get_caps);
   g_object_unref (file);
 
   return TRUE;
 
 error:
+  src->media_type = _NNS_MEDIA_INVALID;
   GST_ERROR_OBJECT (src, "Failed to parse %s", src->json_filename);
   g_free (contents);
   g_object_unref (parser);
-  gst_caps_unref (get_caps);
   g_object_unref (file);
 
   return FALSE;
 }
 
+/**
+ * @brief Setter for datareposrc properties.
+ */
+static void
+gst_data_repo_src_set_property (GObject * object, guint prop_id,
+    const GValue * value, GParamSpec * pspec)
+{
+  GstDataRepoSrc *src;
+  const GstCaps *caps;
+
+  g_return_if_fail (GST_IS_DATA_REPO_SRC (object));
+
+  src = GST_DATA_REPO_SRC (object);
+
+  switch (prop_id) {
+    case PROP_LOCATION:
+      gst_data_repo_src_set_file_path (src, PROP_LOCATION,
+          g_value_get_string (value), NULL);
+      break;
+    case PROP_JSON:
+      gst_data_repo_src_set_file_path (src, PROP_JSON,
+          g_value_get_string (value), NULL);
+      /** To get caps, read JSON before Caps negotiation,
+          to get information on sample data */
+      if (!gst_data_repo_src_read_json_file (src)) {
+        GST_ERROR_OBJECT (src, "Faild to get data format");
+      }
+      break;
+    case PROP_START_SAMPLE_INDEX:
+      src->start_sample_index = g_value_get_uint (value);
+      break;
+    case PROP_STOP_SAMPLE_INDEX:
+      src->stop_sample_index = g_value_get_uint (value);
+      break;
+    case PROP_EPOCHS:
+      src->epochs = g_value_get_uint (value);
+      break;
+    case PROP_IS_SHUFFLE:
+      src->is_shuffle = g_value_get_boolean (value);
+      break;
+    case PROP_TENSORS_SEQUENCE:
+      g_free (src->tensors_seq_str);
+      src->tensors_seq_str = g_value_dup_string (value);
+      if (!gst_data_repo_src_set_tensors_sequence (src)) {
+        GST_ERROR_OBJECT (src, "Faild to set tensors sequence");
+      } else {
+        src->need_changed_caps = TRUE;
+      }
+      break;
+    case PROP_CAPS:
+      caps = gst_value_get_caps (value);
+      if (caps) {
+        src->caps = gst_caps_copy (caps);
```

Comment:
What if src->caps is already set? Doesn't it create a memory leak?

Reply:
I will add the below code. thank you.
```
        new_caps = gst_caps_copy (caps);
        gst_caps_replace (&src->caps, new_caps);
        gst_caps_unref (new_caps);
```